### ISSUE TITLE
Extended markdup UMI support

### ISSUE THREAD

**Issue by Poshi** [samtools/samtools#1710]
UMIs are mostly used to track single molecules and enables us to distinguish them amongst the different reads we got from an experiment. With them, we are able to not mark as duplicates identical reads that come from different molecules.

UMI barcodes are added to the sample in the library preparation step, before PCR; this means that errors can be introduced at the PCR amplification. UMIs are composed of a random sequence of bases (usually of fixed length). As any other base, UMIs are read in the experiment with the same techniques as the DNA we are sequencing, so UMIs are also subject to the same technological read errors.

Most of the errors we expect in UMIs are single base substitutions (either at PCR level or at sequencing level), although some phasing/pre-phasing could also appear.

When deciding how many molecules a set of UMIs represents, many approaches can be taken:
* We can ignore error possibilities and just perform exact matches: if two UMIs match, they are of the same molecule; if they mismatch, they are different molecules.
* We can consider just the number of mismatches: compute the Hamming distance between the two UMIs and use a threshold to decide if they belong to the same molecule or no. How high this threshold should be? I don't know, it depends on chemistry, sequencing technology, UMI length... The longer the UMI, the more possibilities of a read error. Assuming one error in PCR and another in sequencing would make a threshold of 2. Taking into account that UMIs are expected to be random, for a UMI of length 6 having 4 equal bases and two different seems likely to be a couple errors, as I only would expect a couple identical bases in two different molecules.
* We can consider that errors will be scarce and low frequency UMIs will be errors, while common UMIs are the good reads. In that case, given a UMI set, count them and consider the ones that are more frequent than a certain threshold as representing a molecule.
* Other options.

These are just a few examples of how to classify UMIs and molecules. As always happens when processing read errors, we have to guess which was the real data, and there are several ways to do that.
People at [CGAT](http://www.cgat.org/) developed some methods to deal with them and [published their results](https://genome.cshlp.org/content/early/2017/01/18/gr.209601.116.abstract) some time ago. Also, they wrote a nice [blog post](https://cgatoxford.wordpress.com/2015/08/14/unique-molecular-identifiers-the-problem-the-solution-and-the-proof/) where their research is explained. They also developed the [umi_tools](https://github.com/CGATOxford/UMI-tools) to process this kind of data.

Finally, the feature request that required this introduction is easily formulated (not so easily developed): add support in markdup for error handling of UMI errors.
Developing the methods shown in the CGAT paper would be nice, but just adding some kind of error handling would be good.

**Comment by eboyden** [comment]
+1 for this. UMI-aware markdup is much appreciated as it runs substantially faster than Picard or UMI-tools, but currently it severely undercounts duplicates due to the lack of error tolerance.

**Comment by whitwham** [comment]
@Poshi and @eboyden, do either of you have example data that I could look at?  Otherwise I am going to have to make things up which does not always go well.

**Comment by Poshi** [comment]
I have real data (that I probably cannot share and won't be as good as it could be for testing) and some idea on how to generate a sample set.
For the sample set, I would use the sample data that the umi_tools people used in their documentation (UMI barcode, number of reads):
* TCGT -> 2
* CCGT -> 2
* ACGT -> 456
* ACAT -> 72
* AAAT -> 90
* ACAG -> 1

Take a SAM file, keep one aligned read, and for every UMI barcode, create as many copies with that barcode in the RX tag. Finally, modify all the IDs to have different IDs in each read.
With this dataset, you will get different results according to the algorithm you use.
I can generate that file, if you want.

Obviously, using the quality values in BZ/QX (if present) could be useful to refine the results.

Pair end processing should be similar/the same, as the UMI in the RX tag should include all the bases of the entire template separated by hyphens (as the standard says).

Once the group of reads is processed, they should be separated in groups of reads that come from one or more molecules. It would be good to set a couple more of tags:
* MI tag with a molecule identifier (it could be a sequential number)
* Set OX tag to the raw (uncorrected) UMI barcode and RX to the corrected barcode
* Update the quality tags accordingly (BZ (uncorrected) and QX (corrected))

A branch of this addition will be that merge tools will have to take care of keeping MI tags unique in the SAM file, so when merging two or more SAM files you will have to rewrite them (either by keeping them if no duplications or just rewriting them from scratch).

**Comment by daviesrob** [comment]
What sort of workflow do you need this in?  It looks like you could currently [use `UMI tools group`](https://umi-tools.readthedocs.io/en/latest/reference/group.html) to make barcode clusters and write a `BX` tag with the corrected version.  After that's done `samtools markdup --barcode-tag BX` ought to work as the errors have already been accounted for.

Are you trying to avoid the grouping stage, or is there some other reason why you need inexact matches after it's been done?  Or do you need something that can make the clusters faster?

**Comment by Poshi** [comment]
My current workflow uses umi_tools group as you describe. Alignment->Grouping->MarkingDuplicates.

One of the points is that the naïve version of marking duplicates can lead to people using it without realizing the trade-offs if not correcting errors before.
Another point is that a pure C implementation will always be faster than the mostly Python with some C implementation.
Also, it makes sense that the tool that checks duplicates takes into account the possible errors in the UMIs (although we can agree that the errors in the DNA sequence are "corrected" by the alignment step).
Finally, umi_tools group feels painfully slow and memory greedy, taking several times more time than the actual duplicate marking. Having this process done in samtools markdup would speed things up and simplify the pipelines.

So it's not just to avoid a grouping stage, but to integrate it in the tool that will use that grouping; for simplicity and performance reasons. Once the grouping is done, an exact match algorithm can (and should) be used.

**Comment by daviesrob** [comment]
I think UMI grouping is really a two-pass process as you have to find all of the clusters first, and then apply the corrected tags.  As markdup is essentially two-pass already (you have to run `fixmate` first to get the `MC` and `ms` tags it needs) then it might not be too hard to make it discover and apply UMIs in the same workflow.  The first part would need to save its discoveries in a file on the side so they can be applied later.

We'll have a look at some data and UMI tools, to see how easy this might be in practice.

**Comment by eboyden** [comment]
In my experience UMI-tools `group` does not work very well with paired end data - it only tags the first read in a pair, and it doesn't add the duplicate bitflag, which some tools rely on.  (UMI-tools `dedup` works, but I only want to mark duplicates not remove them.)  Whereas Picard does what I want, but using a less sophisticated.  And UMI-tools and Picard are both slow; I'd prefer to replace them entirely with samtools.

**Comment by Poshi** [comment]
@daviesrob I won't say it's a two pass process. You get all the reads for a position and, before deciding on the duplication status, you correct the barcode for that set of reads. Once you ave the corrected UMIs, you decide on the duplication status, set the proper flags and tags and write the results. Then you go to the next aligned position. You don't need to look at all the UMIs of the experiment, only to the set of UMIs that share a position.

@eboyden I completely agree with you. Only the first read of a pair is tagged, which is somehow saving space, but it makes you wonder if the tool that you use for deduplication (samtools, picard...) will work properly or maybe they will fail due to missing tags or, worse, fail silently and spit out wrong results.
Also, umi_tools is not able to correct the UMIs and mark the duplicates in the same run. The dedup tool removes duplicates, the group tool sets the tags. But no easy way to mark the duplicates. Unfortunately, umi_tools feels to me as a half baked pie: they have all the pieces, but it is not fully mounted (lack of functionalities despite having them implemented) nor fully cooked (too slow for production loads). With all my respects with their creators and their useful work.

**Comment by ghuls** [comment]
It might be worth to also check out the rumi implementation: https://github.com/sstadick/rumi

**Comment by whitwham** [comment]
Thanks for al the suggestions.  I am reading them, even though I am not working on markdup at the moment.  I plan to go back to it in the New Year.

**Comment by eboyden** [comment]
Bumping this. Another rust-based tool worth checking out is [rumidup](https://github.com/NKI-GCF/rumidup) (unlike [rumi](https://github.com/sstadick/rumi) it only uses simple hamming distance error correction, but that's still far better than nothing).

**Comment by whitwham** [comment]
Thanks. I am working on it, but more immediate jobs keep getting in the way.

**Comment by eboyden** [comment]
Understood - appreciate the update and the effort.

**Comment by eboyden** [comment]
I discovered another potential issue: read pairs should always be considered together when determining duplicate status or selecting the best representative _pair_ among a pool of duplicates.

I am trying to pipe the output from umi-tools group to samtools markdup, taking advantage of the error correction of the former with the proper bitflag updating of the latter.  But unfortunately umi-tools only tags read 1 in a pair with the consensus UMI (in the BX tag).  This causes samtools to apparently properly mark duplicates for read 1, but seeing no consensus UMI barcode tag in read 2, it marks duplicates based purely on position.  Which leads to data like the following:

```
MN01688:12:000H3MG5N:1:11110:18319:15382        1187    chr1    999212  42      102M    =       999212  102     CTCGGGCGCTGGGGCCTCTGCGGGGGCAGCCGGGGACAGCGCGGCCGGGCGGCGGGAGGGTCCCAGCTGGCGCAGGCAGGCTGCCAGGTGGCCCAGCAGGCG
  /FFFAFFFFFFFFFF=/A/AAFFFFFF/AFFFF/FFAFF/F/FFFFFFAFFFFFFF/FFF/FFA/FF=F=FAF/FFF/FFFF/F//FFF/FFFF//A/FFFF  AS:i:-3 XN:i:0  XM:i:1  XO:i:0  XG:i:0  NM:i:1  MD:Z:41A60      YS:i:-11        YT:Z:CP RG:Z:A    RX:Z:ACTCG-GGGCA        MQ:i:42 MC:Z:102M       ms:i:3447       dt:Z:LB
MN01688:12:000H3MG5N:1:13102:17035:16176        163     chr1    999212  42      102M    =       999212  102     CTCGGGCGCTGGGGCCTCTGCGGGGGCAGCCGGGGACAGCGAGGCCGGGCGGCGGGAGGGTCCCAGCTGGCGCAGGCAGGCTGCCAGGTGGCCCAGCAGGCG
  =FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF6FFFAF/FFFFFFFAFFFFFFFFFFFFFA=FF=FFFFFFAFFFFFFFFFFFFFFFFFF//=FFFFFF  AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:102        YS:i:0  YT:Z:CP RG:Z:A    RX:Z:GCACT-CGATG        MQ:i:42 MC:Z:102M       ms:i:3508       dc:i:2
MN01688:12:000H3MG5N:1:11110:18319:15382        83      chr1    999212  42      102M    =       999212  -102    CTCGGGCGCGGGGGCCGCTGCGGGGGCAGCCGGGGACAGCGAGGCCGGGCGGCGGGAGGGGCCCAGCTGGCGCAGGCAGGCTGCCAGGTGGCCCAGCAGGCG
  F//FFFFA/FFAFFFF/FAFFFF=FFFFFFFFFFFFFFFFF/FFFFFAAFAFAFFFFFFA/FFFFFAAFFFFFAFFAFFFFAF6FFFFAAFFFFFFFFFAFF  AS:i:-11        XN:i:0  XM:i:3  XO:i:0  XG:i:0  NM:i:3  MD:Z:9T6T43T41  YS:i:-3 YT:Z:CP RG:Z:A    RX:Z:ACTCG-GGGCA        MQ:i:42 MC:Z:102M       ms:i:2994       UG:i:42 BX:Z:ACTCGGGGCA dc:i:1
MN01688:12:000H3MG5N:1:13102:17035:16176        83      chr1    999212  42      102M    =       999212  -102    CTCGGGCGCTGGGGCCTCTGCGGGGGCAGCCGGGGACAGCGAGGCCGGGCGGCGGGAGGGTCCCAGCTGGCGCAGGCAGGCTGCCAGGTGGCCCAGCAGGCG
  //FFAAAFA/FFFFF/A/FFFFFFFFFFFFFF/FFFFFFFFFFFFFFFFFFFFAFFFFAFFFFFFFFFFFF=FFFFFFFFFFFFFFFFFFFFFFFFFFFFFF  AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:102        YS:i:0  YT:Z:CP RG:Z:A    RX:Z:GCACT-CGATG        MQ:i:42 MC:Z:102M       ms:i:3591       UG:i:43 BX:Z:GCACTCGATG dc:i:1
```
As you can see, both read 1s are marked unique despite the identical positions, based on different values in the BX tag.  But both read 2s lacks the BX tag, so samtools marks one as unique and the other as a duplicate.  Hence the second read pair is discordant, with read 1 marked as unique and read 2 marked as a duplicate.

It would be best if samtools handled this elegantly - i.e. if the reads have different tag values or one is missing the tag, samtools uses whichever is present or defaults to read 1 if both are present but differ - but barring this, it might be worth making explicit in the documentation that if a barcode tag is specified, both reads in a pair must have it (and have the same value) or else erratic behavior may result.

**Comment by whitwham** [comment]
Can you create a separate issue for that?  That is definitely a bug in the current implementation.

**Comment by eboyden** [comment]
Will do.

**Comment by prosequtor** [comment]
@Poshi @eboyden 
Folks, if you use known/non-random UMI sequences (some library preparation kits have fixed repertoires of UMI sequences), you can run `[fgbio correctumis](https://fulcrumgenomics.github.io/fgbio/tools/latest/CorrectUmis.html)` prior to marking duplicates with `samtools markdup`.

**Comment by Poshi** [comment]
Thanks for the info, @prosequtor, but this is not my case.
